### PR TITLE
Display link text correctly

### LINKED ISSUE #201: Link text displayed incorrectly
The problem with displaying a link appears in the markdown file. To be more precise, once the link text is placed into [] and https or http is included, this link text is followed by the left angle bracket in generated documentation and leads to a non-existing page. In addition, the right angle bracket is added at the end of the reference.

### Expected Behavior

The link text which starts with https or http is correctly displayed in documentation - exactly the same as provided in [], i.e. :
- link text placed in [] in markdown file: `[https://google.com](https://google.com)`
- link text displayed in the documentation: `https://google.com`

### Actual Behavior

The text link displayed in documentation contains extra left angle bracket:

`<https://google.com`

What is more, the reference ends with the right angle bracket and as a result, it redirects to the page which does not exist:

`https://google.com/>`

### Steps to Reproduce

1. Create markdown file i.e. `readme.md` which contains:
```
[https://google.com](https://google.com)
```
2. Create python file i.e. `example.py` which contains:
```
"""
.. include:: ./readme.md
"""
```
3. Run `$ pdoc3 --http localhost:8080 example`
4. Open `$ http://localhost:8080`

image

image

### Additional info

1. Operating system: Ubuntu 19.10
2. pdoc version: [pdoc 0.8.2.dev4+g048919e](https://github.com/pdoc3/pdoc/commit/048919ece483e0609094cee2460d3467d2e0efc3)
3. Python version: 3.8.0

### PR DESCRIPTION
**Pull request by ButterflyBug**
In generated documentation, the link text placed into [] is not followed
by left angle bracket and it redirects to the given website.

Closes #201

### CODE REVIEW

File: pdoc/html_helpers.py [hunk 1]
```diff
@@ -361,9 +361,16 @@ def doctests(text):
     @staticmethod
     def raw_urls(text):
         """Wrap URLs in Python-Markdown-compatible <angle brackets>."""
+        pattern = (
+            r'(?<![<"[(\'])'       # does not start with <, ", \[, \(, '
+            r'(\s*)'               # can contain whitespace
+            r'((?:http|ftp)s?://'  # protocol name
+            r'[^>)\s]+)'           # >, \), whitespace indicate end of the url
```
Comment (kernc) [inline pdoc/html_helpers.py:368]: `\)` won't work for an URL such as `https://en.wikipedia.org/wiki/Orange_(software)`?

File: pdoc/html_helpers.py [hunk 2]
```diff
@@ -361,9 +361,16 @@ def doctests(text):
     @staticmethod
     def raw_urls(text):
         """Wrap URLs in Python-Markdown-compatible <angle brackets>."""
+        pattern = (
+            r'(?<![<"[(\'])'       # does not start with <, ", \[, \(, '
```
Comment (kernc) [inline pdoc/html_helpers.py:365]: `\(` won't work for URLs in unaesthetic yet plausible parentheses: `... find more info elsewhere (https://url).`?

File: pdoc/test/__init__.py [hunk 3]
```diff
@@ -1365,6 +1365,12 @@ def test_urls(self):
         html = to_html(text)
         self.assertEqual(html, expected)
 
+    def test_url_in_link_text(self):
+        text = "[https://google.com](https://google.com)"
```
Comment (kernc) [inline pdoc/test/__init__.py:1369]: Why not add this example into `test_urls()` above, which already tests several cases. Maybe also cover border cases from the other two comments? :sweat_smile:

**Comment by ButterflyBug** [comment]
Hi, I've just updated my PR :smiley: Let me know if you make up some another suspicious type of URLs :smile:

File: pdoc/html_helpers.py [hunk 4]
```diff
@@ -361,9 +361,17 @@ def doctests(text):
     @staticmethod
     def raw_urls(text):
         """Wrap URLs in Python-Markdown-compatible <angle brackets>."""
+        pattern = (
+            r'(?<![<"[\](\'])'                      # does not start with [, <, ", ], (, '
+            r'('
+            r'((?P<surround>\()|(\(?))'             # link can be surrounded with (
+            r'(?P<url>(?:http|ftp)s?://[^>\s]+)'    # url with proocol
+            r'(?(surround)(\))|\)?)'                # if url starts with ( it ends with )
```
Comment (kernc) [inline pdoc/html_helpers.py:369]: Love the commented formatting! :heart_eyes: 

After sleeping on it, I think we can complicate it further. :sweat_smile: Maybe the regex should except only on characters `<"'` as before, and when within markdown link definition, either text part (i.e. somewhere after a non-closed `[`) or URL part (after `](`). That seems like a catch-_less_ approach? 

I'm also certain we should avoid links in inline code, i.e. [non-closed series of `` ` ``](https://daringfireball.net/projects/markdown/syntax#code) backticks. Refs: https://github.com/pdoc3/pdoc/issues/96

File: pdoc/html_helpers.py [hunk 5]
```diff
@@ -361,9 +361,29 @@ def doctests(text):
...
+        pattern = (
+            r'(?P<backtick>`.*`)'                   # finds backticks
+            r'|'
+            r'(?P<markdown_link>(\[.*\])(\(.*\)))'  # gets inline link
+            r'|'
+            r'(?<![<"\'])'                          # does not start with <, ", '
+            r'('
+            r'((?P<left>\()|(\(?))'                 # link can be surrounded with (
+            r'(?P<url>(?:http|ftp)s?://[^>\s]+)'    # url with protocol
+            r'(?(left)(?P<right>\))|\)?)'           # if url starts with (, it ends with )
```
Comment (kernc) [inline pdoc/html_helpers.py:373]: Sorry, I have reservations about this assumption. Besides failing to match
> something something (http://foo and http://bar)

what's it do?

Comment (kernc) [reply pdoc/html_helpers.py:373]: A little different regex seems to pass including this far-fetched test: https://regex101.com/r/CJgYsF/3.
What do you think?

I wonder what regex GitHub/Facebook uses. They seem to work well. :thinking: 
> http://url)
> http://url()
> http://url.
> http://url.url
> http://url,
> http://url,asd

Comment (ButterflyBug) [reply pdoc/html_helpers.py:373]: Test added, your version of regex - checked. All look like working correctly :smile_cat: Thanks a lot for your idea and..another challenge :smile:

File: pdoc/html_helpers.py [hunk 6]
```diff
@@ -361,9 +361,29 @@ def doctests(text):
     @staticmethod
     def raw_urls(text):
         """Wrap URLs in Python-Markdown-compatible <angle brackets>."""
+        pattern = (
+            r'(?P<backtick>`.*`)'               # finds backticks
```
[the remark this answers is not in the document]
Comment (kernc) [reply pdoc/html_helpers.py:365]: Also rename the matching group to `code_span`. Can you also make sure it matches [inline code spans with longer fences](https://spec.commonmark.org/0.29/#code-spans)? This test, basically:
```markdown
lala ``x ` http://foo`` http://bar `http://foo`
```

File: pdoc/html_helpers.py [hunk 7]
```diff
@@ -361,9 +361,29 @@ def doctests(text):
...
+            r')'
+        )
+
+        def replace(match):
+            groups = match.groupdict("")
+
+            if groups["backtick"] or groups["markdown_link"]:
+                return match.group()
+            else:
+                return "<" + groups["url"] + ">"
```
Comment (kernc) [inline pdoc/html_helpers.py:383]: Also maybe invert this `if` to check for `url` group first? I don't like `else` on short-circuit returns. This now almost fits on one line!

File: pdoc/test/__init__.py [hunk 8]
```diff
@@ -1352,7 +1352,16 @@ def test_urls(self):
...
 http://url.com
-```"""
+```
+[https://google.com](https://google.com)
+[https://en.wikipedia.org/wiki/Orange_(software)](https://en.wikipedia.org/wiki/Orange_(software))
+https://en.wikipedia.org/wiki/Orange_(software)
+(https://google.com)
+[Check https://google.com here](https://google.com)
+`https://google.com`
+(http://foo and http://bar)
```
Comment (kernc) [inline pdoc/test/__init__.py:1362]: Finally, maybe split the tests by matching and non-matching together. Yea, I'm pushing it. :sweat_smile:

File: pdoc/html_helpers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ButterflyBug) [reply pdoc/html_helpers.py:365]: haha no worries ;D I learn a lot from these whims. Besides, thanks to them our regex looks better and better. :smile: I agree that using `re.VERBOSE`   would be a prettier solution, thanks for that comment :smile:

File: pdoc/test/__init__.py (hunk 8 above)
Comment (ButterflyBug) [reply pdoc/test/__init__.py:1362]: I'm not sure if I understood that correctly. :thinking:  You mean to split `(http://foo and http://bar)` into two separated tests like `(http://foo something)` and `(something http//:bar)?` :smile:

Comment (kernc) [reply pdoc/test/__init__.py:1362]: I meant to split all these test lines into those that match (and should produce `<url>`) and those that don't (shouldn't).

**Comment by kernc** [comment]
I pushed some pedantic touch-ups I didn't feel like voicing aloud. This now looks good and should hopefully stick as far as wrapping raw URLs goes. :crossed_fingers: Much appreciated!

If you care to make other regexes in the module more reader-friendly, or if you find other bugs, obviously welcome. :grin: Thanks!

**Pull request merged**